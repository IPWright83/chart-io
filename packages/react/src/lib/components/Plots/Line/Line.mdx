import { Meta, Story, Canvas } from "@storybook/addon-docs/blocks";

# Line Plots

There are two different Line based components, these include:

-   `<Line>` - allows us to define a line plot on our chart
-   `<Lines>` - allows us to define multiple line series

> **Note**: To ensure interactivity this component should be used with the `<EventRecorder>` component which picks up events on the background and pipes them to the plot.

## `<Line>` Component

The `<Line>` component should be used when you want to plot a single line plot on a chart.

<Canvas>
    <Story id="xycharts-line--basic" />
</Canvas>

### Props

| Prop          | Type      | Default            | Note                                                                                                         |
| ------------- | --------- | ------------------ | ------------------------------------------------------------------------------------------------------------ |
| `color`       | `string`  | Derived from Theme | Override the colour of the series                                                                            |
| **`x`\***     | `string`  | null               | The key of the field that contains the value along the x-axis.                                               |
| **`y`\***     | `string`  | null               | The key of the field that contains the value along the y-axis.                                               |
| `interactive` | `boolean` | true               | Whether the plot should be interactive. Setting this to false will disable tooltips & markers from this plot |

### Visual Styling

Most of the configuration for a `<Line>` is similar to other plots. Currently the only styling property avaliable is the colour option.

<Canvas>
    <Story id="xycharts-scatter--color" />
</Canvas>

## `<Lines> Component`

The `<Lines>` component allows us to define multiple line series. When plotting multiple line series they do not stack, but always behave by overlaying on top of one another.

### Props

| Prop       | Type       | Default            | Note                                                           |
| ---------- | ---------- | ------------------ | -------------------------------------------------------------- |
| `colors`   | `string[]` | Derived from Theme | Override the colours for the series                            |
| **`x`\***  | `string`   | null               | The key of the field that contains the value along the x-axis. |
| **`ys`\*** | `string[]` | null               | The keys of the fields that contains values along the y-axis.  |

### Overlapping Lines

The simplest way is to use the `<Lines>` (plural) component. The `<Lines>` component allows us to define multiple series at once and takes care of automaticallying layering and assigning colors. It is however possible to do this manually if required by defining multiple `<Line>` components, configuring with a colour manually.

The results of either approach, render in exactly the same way:

<Canvas>
    <Story id="xycharts-area--multiple-areas" />
</Canvas>
