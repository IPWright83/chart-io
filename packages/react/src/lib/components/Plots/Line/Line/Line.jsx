import PropTypes from "prop-types";
import React from "react";

import { withCanvas, withSVG, withXYPlot } from "../../../../hoc";

import { CanvasLine } from "./CanvasLine";
import { SVGLine } from "./SVGLine";

const WrappedCanvasLine = withCanvas(withXYPlot(CanvasLine), "plot line");
const WrappedSVGLine = withSVG(withXYPlot(SVGLine), "plot line");

/**
 * Represents a Line plot
 * @param  {Object} props       The set of React properties
 * @return {ReactDOMComponent}  The Line plot component
 */
const Line = ({ useCanvas, ...props }) => {
    if (useCanvas) {
        return <WrappedCanvasLine {...props} />;
    }

    return <WrappedSVGLine {...props} />;
};

Line.propTypes = {
    ...SVGLine.propTypes,

    /**
     * Should this plot use an HTML Canvas instead of SVG?
     * @type {[type]}
     */
    useCanvas: PropTypes.bool,
};

Line.requiresVirtualCanvas = false;

export { Line };
