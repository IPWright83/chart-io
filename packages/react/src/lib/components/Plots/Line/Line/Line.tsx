import type { IPlotProps } from "@chart-io/types";
import React from "react";

import { withCanvas, withSVG, withXYPlot } from "../../../../hoc";

import { CanvasLine } from "./CanvasLine";
import { SVGLine } from "./SVGLine";

export interface ILineProps extends Omit<IPlotProps, "layer" | "canvas"> {
    /**
     * Should Canvas be used instead of SVG?
     */
    useCanvas?: boolean;
}

const WrappedCanvasLine = withCanvas(withXYPlot<ILineProps>(CanvasLine), "plot line");
const WrappedSVGLine = withSVG(withXYPlot<ILineProps>(SVGLine), "plot line");

/**
 * Represents a Line plot
 * @param  props       The set of React properties
 * @return             The Line plot component
 */
export function Line({ useCanvas = false, ...props }: ILineProps) {
    if (useCanvas) {
        return <WrappedCanvasLine {...props} />;
    }

    return <WrappedSVGLine {...props} />;
}

Line.requiresVirtualCanvas = false;
Line.isPlot = true;
Line.brush = {
    horizontal: true,
    vertical: false,
};
Line.zoom = {
    horizontal: true,
    vertical: false,
};
