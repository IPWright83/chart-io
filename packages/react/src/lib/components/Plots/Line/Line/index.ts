export * from "./Line";
export * from "./CanvasLine";
export * from "./SVGLine";
