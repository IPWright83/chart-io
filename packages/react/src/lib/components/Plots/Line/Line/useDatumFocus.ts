import * as d3 from "@chart-io/d3";
import type { IColor, ICoordinate, IMouseEventType, IScale } from "@chart-io/types";
import { useEffect } from "react";

import { eventActions, IDispatch } from "../../../../store";
import { isNullOrUndefined } from "../../../../utils";

/**
 * Responds to events from an event layer to focus the nearest datum
 * @param  dispatch   The redux store dispatch function
 * @param  layer      The layer we're updating to from a React useRef
 * @param  x          The key of the field used for the x position
 * @param  y          The key of the field used for the y position
 * @param  xScale     The d3 scale for the x axis
 * @param  yScale     The d3 scale for the y axis
 * @param  data       The complete dataset for the plot
 * @param  eventMode  The event mode from the selector one of ["NONE", "ENTER", "MOVE"]
 * @param  position   The { x, y } position of the mouse
 * @param  color      The color of the plot
 */
export function useDatumFocus(
    dispatch: IDispatch,
    layer: React.MutableRefObject<Element>,
    x: string,
    y: string,
    xScale: IScale | undefined,
    yScale: IScale | undefined,
    data,
    eventMode: IMouseEventType,
    position: ICoordinate,
    color: IColor
) {
    /* If possible respond to global mouse events for tooltips etc */
    useEffect(() => {
        if (!layer.current || !xScale || !yScale || !data) {
            // istanbul ignore next
            return;
        }

        // We need the invert function to be able to do this
        // @ts-expect-error: We're handling a runtime validation so this is fine
        if (!xScale.invert) {
            // istanbul ignore next
            return;
        }

        // Don't run for the EXIT event
        if (eventMode === "NONE") {
            d3.select(layer.current).selectAll("circle").remove();
            return;
        }

        // Work out the datum that we're closet to
        // @ts-expect-error: We've already aserted that invert exists
        const xValue = xScale.invert(position.x);
        const index = d3.bisector((d) => d[x]).center(data, xValue);
        const datum = data[index];

        if (isNullOrUndefined(datum[y])) {
            // istanbul ignore next
            return;
        }

        // Get the appropriate attributes
        const cx = xScale(datum[x]);
        const cy = yScale(datum[y]);
        const fill = color;

        const marker = { fill, cx, cy };
        const horizontalDropline = {
            isHorizontal: true,
            color: fill,
            x1: cx,
            x2: xScale.range()[0],
            y1: cy,
            y2: cy,
        };
        const verticalDropline = {
            isVertical: true,
            color: fill,
            x1: cx,
            x2: cx,
            y1: cy,
            y2: yScale.range()[0],
        };

        dispatch(eventActions.addMarker(marker));
        dispatch(eventActions.addDropline(horizontalDropline));
        dispatch(eventActions.addDropline(verticalDropline));

        // Clean up operations on exit
        return () => {
            dispatch(eventActions.removeMarker(marker));
            dispatch(eventActions.removeDropline(horizontalDropline));
            dispatch(eventActions.removeDropline(verticalDropline));
        };
    }, [xScale, yScale, x, y, position.x, position.y, color, data, dispatch, eventMode, layer]);
}
