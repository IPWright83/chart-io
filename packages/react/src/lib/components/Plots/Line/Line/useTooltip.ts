import * as d3 from "d3";
import { useEffect } from "react";

import { eventActions } from "../../../../store";
import { isNullOrUndefined } from "../../../../utils";

/**
 * Responds to events from an event layer to show Tooltips
 * @param  {Function} dispatch The redux store dispatch function
 * @param  {Object} layer      The layer we're updating to from a React useRef
 * @param  {String} x          The key of the field used for the x position
 * @param  {String} y          The key of the field used for the y position
 * @param  {d3.Scale} xScale   The d3 scale for the x axis
 * @param  {d3.Scale} yScale   The d3 scale for the y axis
 * @param  {Object[]} data        The complete dataset for the plot
 * @param  {String} eventMode  The event mode from the selector one of ["NONE", "ENTER", "MOVE"]
 * @param  {Object} position   The { x, y } position of the mouse
 * @param  {String} color      The color of the plot
 */
const useTooltip = (dispatch, layer, x, y, xScale, yScale, data, eventMode, position, color) => {
    /* If possible respond to global mouse events for tooltips etc */
    useEffect(() => {
        if (!layer.current || !xScale || !yScale || !data) {
            // istanbul ignore next
            return;
        }

        // We need the invert function to be able to do this
        if (!xScale.invert) {
            // istanbul ignore next
            return;
        }

        // Don't run for the EXIT event
        if (eventMode === "NONE") {
            d3.select(layer.current).selectAll("circle").remove();
            return;
        }

        // Work out the datum that we're closet to
        const xValue = xScale.invert(position.x);
        const index = d3.bisector((d) => d[x]).center(data, xValue);
        const datum = data[index];

        if (isNullOrUndefined(datum[y])) {
            // istanbul ignore next
            return;
        }

        // Common x value
        const tooltipItemX = {
            datum,
            name: x,
            value: datum[x],
        };
        dispatch(eventActions.addTooltipItem(tooltipItemX));

        const tooltipItemY = {
            datum,
            name: y,
            value: datum[y],
            seriesType: "column",
            fill: color,
        };
        dispatch(eventActions.addTooltipItem(tooltipItemY));
        dispatch(eventActions.setPositionEvent(position.x, position.y));

        // Clean up operations on exit
        return () => {
            dispatch(eventActions.setTooltipBorderColor(undefined));
            dispatch(eventActions.removeTooltipItem(tooltipItemX));
            dispatch(eventActions.removeTooltipItem(tooltipItemY));
        };
    }, [eventMode, position.x, position.y, xScale, yScale, x, y, data, layer, color]);
};

export { useTooltip };
