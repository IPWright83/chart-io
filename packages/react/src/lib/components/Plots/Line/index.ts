export * from "./Line";
export * from "./Lines";
