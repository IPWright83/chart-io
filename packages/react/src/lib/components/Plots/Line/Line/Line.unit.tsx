import * as d3 from "d3";
import React from "react";
import { Provider } from "react-redux";
import { render, act } from "@testing-library/react";
import { toMatchImageSnapshot } from "jest-image-snapshot";

import { VirtualCanvas, VIRTUAL_CANVAS_DEBOUNCE } from "../../../VirtualCanvas";
import { Line } from "./Line";
import { createStore } from "../../../../store";

expect.extend({ toMatchImageSnapshot });

import { getBuffer, wait, renderChart } from "../../../../testUtils";

describe("Line", () => {
    const data = [
        { y: 0, x: 0 },
        { y: 5, x: 1 },
        { y: 10, x: 2 },
        { y: 0, x: 3 },
    ];

    const scales = {
        y: d3.scaleLinear().domain([0, 20]).range([100, 0]),
        x: d3.scaleLinear().domain([0, 5]).range([0, 100]),
    };

    describe("using SVG", () => {
        it("should render correctly", async () => {
            const { asFragment } = await renderChart({
                children: <Line x="x" y="y" />,
                data,
                scales,
            });

            // Wait for the second render of the line, as
            // first render we put in a placeholder to animate
            await wait(10);

            expect(asFragment()).toMatchSnapshot();
        });

        describe("should skip rendering if", () => {
            it("there is no x scale avaliable", async () => {
                const { asFragment } = await renderChart({
                    children: <Line x="x" y="y" />,
                    data,
                    scales: { y: scales.y },
                });

                expect(asFragment()).toMatchSnapshot();
            });

            it("there is no y scale avaliable", async () => {
                const { asFragment } = await renderChart({
                    children: <Line x="x" y="y" />,
                    data,
                    scales: { x: scales.x },
                });

                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe("should respond to event", () => {
            it("finding nearest datum and highlight the point correctly", async () => {
                const store = createStore();

                render(
                    <Provider store={store}>
                        <svg>
                            <Line x="x" y="y" />
                        </svg>
                    </Provider>
                );

                // Wait for the second render of the line, as
                // first render we put in a placeholder to animate
                await wait(10);

                act(() => {
                    store.dispatch({ type: "CHART.SET_SCALES", payload: { fields: ["x"], scale: scales.x } });
                    store.dispatch({ type: "CHART.SET_SCALES", payload: { fields: ["y"], scale: scales.y } });
                    store.dispatch({ type: "CHART.SET_DATA", payload: data });
                    store.dispatch({ type: "CHART.SET_DIMENSIONS", payload: { width: 200, height: 200 } });
                });

                // Spy on the store for the updates from the Area chart
                const originalDispatch = store.dispatch;
                jest.spyOn(store, "dispatch").mockImplementationOnce((action) => originalDispatch(action));

                act(() => {
                    // Simulate a mouse move on the background
                    store.dispatch({ type: "EVENT.MOUSE_ENTER", payload: { offsetX: 25, offsetY: 80 } });
                });

                await wait(1);

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);
                expect(dispatchCalls).toEqual([
                    "EVENT.MOUSE_ENTER",
                    "EVENT.ADD_MARKER",
                    "EVENT.ADD_DROPLINE",
                    "EVENT.ADD_DROPLINE",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                ]);
            });
        });
    });

    describe("using Canvas", () => {
        it("should render correctly", async () => {
            const { container } = await renderChart({
                children: (
                    <VirtualCanvas>
                        <Line x="x" y="y" useCanvas={true} />
                    </VirtualCanvas>
                ),
                data,
                scales,
            });

            // Wait for the second render of the line, as
            // first render we put in a placeholder to animate
            await wait(10);

            await wait(VIRTUAL_CANVAS_DEBOUNCE * 2);

            const canvasBuffer = getBuffer(container.querySelector(".canvas"));
            expect(canvasBuffer).toMatchImageSnapshot();

            // const virtualCanvasBuffer = getBuffer(container.querySelector(".virtual-canvas"));
            // expect(virtualCanvasBuffer).toMatchImageSnapshot();
        });

        describe("should respond to event", () => {
            it("finding nearest datum and highlight the point correctly", async () => {
                const store = createStore();

                render(
                    <Provider store={store}>
                        <svg>
                            <VirtualCanvas>
                                <Line x="x" y="y" useCanvas={true} />
                            </VirtualCanvas>
                        </svg>
                    </Provider>
                );

                // Wait for the second render of the line, as
                // first render we put in a placeholder to animate
                await wait(10);

                act(() => {
                    store.dispatch({ type: "CHART.SET_SCALES", payload: { fields: ["x"], scale: scales.x } });
                    store.dispatch({ type: "CHART.SET_SCALES", payload: { fields: ["y"], scale: scales.y } });
                    store.dispatch({ type: "CHART.SET_DATA", payload: data });
                    store.dispatch({ type: "CHART.SET_DIMENSIONS", payload: { width: 200, height: 200 } });
                });

                // Spy on the store for the updates from the Area chart
                const originalDispatch = store.dispatch;
                jest.spyOn(store, "dispatch").mockImplementationOnce((action) => originalDispatch(action));

                // Simulate a mouse move on the background
                act(() => {
                    store.dispatch({ type: "EVENT.MOUSE_ENTER", payload: { offsetX: 25, offsetY: 80 } });
                });

                await wait(1);

                const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);
                expect(dispatchCalls).toEqual([
                    "EVENT.MOUSE_ENTER",
                    "EVENT.ADD_MARKER",
                    "EVENT.ADD_DROPLINE",
                    "EVENT.ADD_DROPLINE",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.ADD_TOOLTIP_ITEM",
                    "EVENT.SET_POSITION_TOOLTIP_ITEM_EVENT",
                ]);
            });
        });
    });
});
