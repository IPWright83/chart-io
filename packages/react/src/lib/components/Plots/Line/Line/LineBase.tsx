import { chartSelectors, IState, line } from "@chart-io/core";
import { d3 } from "@chart-io/core";
import type { IPlotProps } from "@chart-io/types";

import { useMemo } from "react";
import { useSelector } from "react-redux";

import { useLegendItem, useRender } from "../../../../hooks";

import { useDatumFocus } from "./useDatumFocus";
import { usePathCreator } from "./usePathCreator";
import { useTooltip } from "./useTooltip";

export function LineBase({
    x,
    y,
    color,
    scaleMode = "plot",
    showInLegend = true,
    interactive = true,
    layer,
    canvas,
}: IPlotProps) {
    const data = useSelector((s: IState) => chartSelectors.data(s));
    const xScale = useSelector((s: IState) => chartSelectors.scales.getScale(s, x, scaleMode));
    const yScale = useSelector((s: IState) => chartSelectors.scales.getScale(s, y, scaleMode));
    const theme = useSelector((s: IState) => chartSelectors.theme(s));
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));
    const animationDuration = useSelector((s: IState) => chartSelectors.animationDuration(s));

    const seriesColor = color || theme.series.colors[0];
    const sortedData = useMemo(() => data.toSorted((a, b) => d3.ascending(a[x], b[x])), [data, x]);

    useLegendItem(y, "line", showInLegend, seriesColor);

    // Used to create our initial path
    usePathCreator(layer, x, y, xScale, yScale);

    /* On future renders we want to update the path */
    useRender(() => {
        const props = { x, y, xScale, yScale, data: sortedData, color: seriesColor, theme };

        // Handle Canvas rendering
        if (canvas) {
            return line.canvas.render({ ...props, height, width, canvas });
        }

        line.render({ ...props, layer: layer.current, animationDuration });
    }, [x, y, sortedData, xScale, yScale, width, height, seriesColor, theme, layer, canvas, animationDuration]);

    // If possible respond to global mouse events for tooltips etc
    useDatumFocus({ layer, x, y, interactive, xScale, yScale, data: sortedData, color: seriesColor });
    useTooltip({ layer, x, y, interactive, xScale, yScale, data: sortedData, color: seriesColor });
}
