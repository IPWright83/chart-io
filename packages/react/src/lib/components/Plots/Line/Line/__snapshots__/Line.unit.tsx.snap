// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line using SVG should render correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot line"
      clip-path="url(#clip-path-test)"
    >
      <path
        class="line"
        d="M0,100L20,75L40,50L60,100                                                "
        style="fill: none; pointer-events: none; stroke: rgb(153, 193, 220); stroke-width: 2px;"
      />
      <path
        class="line"
        d="M0,100L100,100"
        style="fill: none; pointer-events: none;"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Line using SVG should skip rendering if there is no x scale avaliable 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot line"
      clip-path="url(#clip-path-test)"
    />
  </svg>
</DocumentFragment>
`;

exports[`Line using SVG should skip rendering if there is no y scale avaliable 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-plot line"
      clip-path="url(#clip-path-test)"
    />
  </svg>
</DocumentFragment>
`;
